### PR TITLE
Ensure container ports without names are also included in the head node service

### LINKED ISSUE #463: [Bug] [Operator] [Controller] [Service] | Ports without name are ignored
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

## Issue

If we create a new `rayclusters.ray.io` and we set in the headnode the following options
```
  ports:
  - containerPort: 10001
    protocol: TCP
  - containerPort: 8265
    protocol: TCP
  - containerPort: 8000
    protocol: TCP
  - containerPort: 6379
    protocol: TCP
  - containerPort: 9001
    protocol: TCP
```

The ray operator will create the deployment but it will not include in the headnode kubernetes service the above ports. In turn, this leads to workers being unable to communicate with the head node and timing out trying to connect to a non-existing port.

If however we use a port names, then the service will be updated with the right ports

## What should happen

The crd should validate that the container ports have a valid name, or the operator should reject crd and throw an error instead of spawning the pods.

### Reproduction script

Create or edit a `rayclusters.ray.io`  and add in the headnode some ports without a name, e.g.

```
  ports:
  - containerPort: 10001
    protocol: TCP
  - containerPort: 8265
    protocol: TCP
  - containerPort: 8000
    protocol: TCP
  - containerPort: 6379
    protocol: TCP
  - containerPort: 9001
    protocol: TCP
```

The cluster will appear to be ok, but what will actually happen is workers starting and terminating after timing out

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yicheng-Lu-llll**
## Why are these changes needed?

[issue 463](https://github.com/ray-project/kuberay/issues/463) describes that if container port names are not set in the ray cluster, workers are unable to communicate with the head due to the wrong head node Kubernetes service config.
This is because in the current [implementation](https://github.com/ray-project/kuberay/blob/dbcc686eabefecc3b939cd5c6e7a051f2473ad34/ray-operator/controllers/ray/common/service.go#L183):

```go
	for _, port := range cPorts {
		svcPorts[port.Name] = port.ContainerPort
	}
```
if port.Name == "", Only the last container port will be considered since all container ports have identical keys (port.Name) .

Considering these:

1. container port names are optional ([link](https://pkg.go.dev/k8s.io/api/core/v1#ContainerPort)). -> Completely legal to not set or partially set container port names.

2. multi-port-services must set port names ([link](https://pkg.go.dev/k8s.io/api/core/v1#ServicePort) [link](https://kubernetes.io/docs/concepts/services-networking/service/#multi-port-services)). -> head node Kubernetes service must set service port names.

3. As long as service port numbers are equal to container port numbers, the head node service's functionalities will work just fine( e. g. workers can communicate with the head). -> There is no extra constraint to the service port name or container port name except the name should be unique. -> This means though it is great to set the service port name equal to the container port name for better understanding, it is not mandatory.

4. Fact 1,2,3 -> The fix should keep all functionalities working even without setting container port names. It can be done by giving service port names.

5. Gcs, dashboard, and client's Port number can be changed([link](https://docs.ray.io/en/latest/ray-core/configure.html#head-node)). ->  We cannot hardcode the service port names since we have no idea what the associate port number is used for.

6. Fact 4,5 -> Assigning random but unique service port names if the user does not set the container port names is a good idea.

With the above consideration, This PR adds the following to fix the [issue 463](https://github.com/ray-project/kuberay/issues/463):
```go
	for _, port := range cPorts {
		if port.Name == "" {
			port.Name = fmt.Sprint(port.ContainerPort) + "-port"
		}
		svcPorts[port.Name] = port.ContainerPort
	}
```

## Related issue number
Closes #463

## Checks
Step1: 
Clone my repo and checkout to `PortsWithoutNameAreIgnored` branch

Step2: 
Follow [DEVELOPMENT.md ](https://github.com/ray-project/kuberay/blob/master/ray-operator/DEVELOPMENT.md)to build Docker image and install kuberay operator.

 Step3: 
 remove container port names in kuberay/ray-operator/config/samples/ray-cluster.complete.yaml:

 ```
            ports:
          - containerPort: 6379
           
          - containerPort: 8265
          
          - containerPort: 10001
         
```

Step4: 
```sh
#path: kuberay/
kubectl apply -f ray-operator/config/samples/ray-cluster.complete.yaml
```

Step5: 
```sh
kubectl describe svc raycluster-complete-head-svc
#should see:
#Port:              8265-port  8265/TCP
#TargetPort:        8265/TCP
#Endpoints:         <none>
#Port:              10001-port  10001/TCP
#TargetPort:        10001/TCP
#Endpoints:         <none>
#Port:              6379-port  6379/TCP
#TargetPort:        6379/TCP
#Endpoints:         <none>
```

### CODE REVIEW

**Review by kevin85421 — COMMENTED** [review]
Thanks @Yicheng-Lu-llll for this contribution! Do you mind add some tests?

**Comment by kevin85421** [comment]
Can you rebase with master? #893 can fix the CI failure in "Go-build-and-test / Build Operator Binaries and Docker Images". Thanks!

**Comment by Yicheng-Lu-llll** [comment]
Updated

**Review by kevin85421 — APPROVED** [review]
LGTM

File: ray-operator/controllers/ray/common/service_test.go [hunk 1]
```diff
@@ -126,3 +137,27 @@ func TestBuildServiceForHeadPodWithAnnotations(t *testing.T) {
 		t.Fatalf("Expected `%v` but got `%v`", annotations, svc.ObjectMeta.Annotations)
 	}
 }
+
+func TestGetPortsFromCluster(t *testing.T) {
+	svcPorts, err := getPortsFromCluster(*instanceWithWrongSvc)
+	assert.Nil(t, err)
+
+	svcNames := map[int32]string{}
+	for name, port := range svcPorts {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/service_test.go:146]: Add some comments.

**Review by kevin85421 — APPROVED** [review]
I also tested this PR manually.

Screen Shot 2023-02-14 at 11 56 02 AM

**Pull request merged**